Makefile

```diff
@@ -36,98 +59,166 @@ else
+CFLAGS := -G 0 -c -nostdinc -march=vr4300 -mfix4300 -mabi=32 -mips3 -mno-abicalls -mdivide-breaks -fno-PIC -fno-common -ffreestanding -fbuiltin -fno-builtin-sinf -fno-builtin-cosf -funsigned-char $(WARNINGS)
 CFLAGS += -fno-strict-aliasing # TODO: Try adjusting code to remove this
-ASFLAGS := -w -nostdinc -c -G 0 -march=vr4300 -mabi=32 -mgp32 -mfp32 -DMIPSEB -D_LANGUAGE_ASSEMBLY -D_MIPS_SIM=1 -D_ULTRA64
-CPPFLAGS = -DMODERN_CC -D_MIPS_SZLONG=32 -D__USE_ISOC99 $(GBIDEFINE) $(VERSION_DEFINE) $(DEBUGFLAG)
-IINC = -I . -I $(WORKING_DIR)/include -I $(WORKING_DIR)/include/compiler/modern_gcc -I $(WORKING_DIR)/include/PR
-MIPS_VERSION := -mips3
-ASOPTFLAGS :=
-
-ifeq ($(findstring _d,$(TARGET)),_d)
-OPTFLAGS := -Og -ggdb3 -ffast-math -fno-unsafe-math-optimizations
-else
-OPTFLAGS := -Os -ggdb3 -ffast-math -fno-unsafe-math-optimizations
-endif
+CFLAGS += -D_MIPS_SZLONG=32 -D__USE_ISOC99 $(C_DEFINES) $(DEF_INC_CFLAGS)
 
-ifeq ($(findstring _rom,$(TARGET)),_rom)
-CPPFLAGS += -D_FINALROM
-endif
+# C preprocessor flags
+CPPFLAGS := -P -Wno-trigraphs $(DEF_INC_CFLAGS)
 
-SRC_DIRS := $(shell find src -type d)
-C_FILES  := $(foreach dir,$(SRC_DIRS),$(wildcard $(dir)/*.c))
-S_FILES  := $(foreach dir,$(SRC_DIRS),$(wildcard $(dir)/*.s))
+# tools
+PRINT = printf
 
-# Versions J and below used the C matrix math implementations
-MGU_MATRIX_FILES := mtxcatf normalize scale translate
-ifeq ($(MGU_ASM), 1)
-C_FILES := $(filter-out $(addprefix src/gu/,$(MGU_MATRIX_FILES:=.c)),$(C_FILES))
-else
-S_FILES := $(filter-out $(addprefix src/mgu/,$(MGU_MATRIX_FILES:=.s)),$(S_FILES))
+ifeq ($(COLOR),1)
+NO_COL  := \033[0m
+RED     := \033[0;31m
+GREEN   := \033[0;32m
+BLUE    := \033[0;34m
+YELLOW  := \033[0;33m
+BLINK   := \033[33;5m
 endif
 
-C_O_FILES := $(foreach f,$(C_FILES:.c=.o),$(BUILD_DIR)/$f)
-S_O_FILES := $(foreach f,$(S_FILES:.s=.o),$(BUILD_DIR)/$f)
-O_FILES   := $(S_O_FILES) $(C_O_FILES)
+# Common build print status function
+define print
+  @$(PRINT) "$(GREEN)$(1) $(YELLOW)$(2)$(GREEN) -> $(BLUE)$(3)$(NO_COL)\n"
+endef
 
-$(shell mkdir -p src $(foreach dir,$(SRC_DIRS),$(BUILD_DIR)/$(dir)))
+#==============================================================================#
+# Main Targets                                                                 #
+#==============================================================================#
 
-.PHONY: all clean distclean setup
-all: $(BUILD_AR)
-
-$(BUILD_AR): $(O_FILES)
-	@printf "    [AR] $@\n"
-	$(V)$(AR) rcs $@ $^
+# Default target
+default: $(LIB)
 
 clean:
-	$(RM) -rf $(BUILD_DIR)
+	$(RM) -r $(BUILD_DIR_BASE)
 
-distclean:
-	$(RM) -rf extracted/ $(BUILD_ROOT)
+ALL_DIRS := $(BUILD_DIR) $(addprefix $(BUILD_DIR)/,$(SRC_DIRS))
 
-GBIDEFINE := -DF3DEX_GBI
+# Make sure build directory exists before compiling anything
+DUMMY != mkdir -p $(ALL_DIRS)
```

Comment:
What's the difference between `DUMMY !=mkdir -p $(ALL_DIRS)` and `$(shell mkdir -p $(ALL_DIRS))`? I find the former less clear to read

Reply:
IDK that's been there for years, I never really understood it